packages/itwinui-react/src/core/utils/hooks/useSyncExternalStore.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import * as React from 'react';
+const _React = React; // prevent bundlers from stripping the namespace import
+
+/**
+ * Wrapper around `React.useSyncExternalStore` that uses a shim for React 17.
+ */
+export const useSyncExternalStore =
+  _React.useSyncExternalStore || useSyncExternalStoreShim;
+
+// ----------------------------------------------------------------------------
+
+/**
+ * The shim below is adapted from React's source to make it ESM-compatible.
+ *
+ * Note: This does not use `getServerSnapshot` at all, because there is
+ * apparently no way to check "hydrating" state in pre-18.
+ *
+ * @see https://github.com/facebook/react/tree/main/packages/use-sync-external-store
+ */
+function useSyncExternalStoreShim<T>(
+  subscribe: (onSubscribe: () => void) => () => void,
+  getSnapshot: () => T,
+): T {
+  const value = getSnapshot();
+  const [{ instance }, forceUpdate] = React.useState({
+    instance: { value, getSnapshot },
+  });
+
+  React.useLayoutEffect(() => {
```

Comment:
I feel comments like [this](https://redirect.github.com/facebook/react/blob/2b036d3f1f016fbe8b121d223d96f09e785b97e1/packages/use-sync-external-store/src/useSyncExternalStoreShimClient.js#L104-L105) and [this](https://redirect.github.com/facebook/react/blob/2b036d3f1f016fbe8b121d223d96f09e785b97e1/packages/use-sync-external-store/src/useSyncExternalStoreShimClient.js#L116-L117) make it easy to understand the code. So, it would be helpful to have them here too.

Reply:
i deliberately removed the comments because it's not code that will ever be touched. it's write-once-and-forget.

as mentioned in the PR description, i would have liked to use the `use-sync-external-store` npm package, but react doesn't publish an ESM version of it